src/Data/Array/Mutable/Linear.hs

```diff
@@ -11,75 +15,123 @@
 --   - TODO:
 --   - Add more of the API from Data.Vector
 --   - Remove the use of error on indicies out of bound.
+--   - Use toVec instead of toList
+--   - Read should return an (Unrestricted a)
 module Data.Array.Mutable.Linear
   ( Array,
     alloc,
     write,
     read,
     length,
     resize,
-    resizeSeed
+    resizeSeed,
+    toList,
+    fromList
   )
 where
 
-import Data.Unrestricted.Linear (Unrestricted (..))
-import GHC.Exts
+import Data.Unrestricted.Linear
+import GHC.Exts hiding (toList, fromList)
 import GHC.Stack
-import qualified Unsafe.Linear as Linear
+import qualified Unsafe.Linear as Unsafe
+import qualified Unsafe.MutableArray as Unsafe
 import Prelude hiding (length, read)
-import Unsafe.MutableArray
+import qualified Prelude
 
 -- # Data types
-----------------------------
+-------------------------------------------------------------------------------
 
 data Array a where
   Array :: Int -> (MutableArray# RealWorld a) -> Array a
 
 -- # Creation
-----------------------------
+-------------------------------------------------------------------------------
 
 -- | Allocate a constant array given a size and an initial value
-alloc :: Int -> a -> (Array a #-> Unrestricted b) -> Unrestricted b
-alloc size x f = f (Array size (newMutArr size x))
+-- Note that this does throw an exception and that there's a reasonable
+-- coherent view and method to using exceptions in Haskell!
+-- See https://www.tweag.io/posts/2020-04-16-exceptions-in-haskell.html.
```

Comment:
No for this PR, but this is not very helpful. You won't be repeating this on every partial function. Simply specify your precondition, instead.